### ISSUE TITLE
completions: Support adding additional code to complete values

### ISSUE THREAD

**Issue by epage** [epage/clapng#65]
<a href="https://github.com/joshtriplett"><img src="https://avatars.githubusercontent.com/u/162737?v=4" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [joshtriplett](https://github.com/joshtriplett)**
_Monday Jul 04, 2016 at 12:48 GMT_
_Originally opened as https://github.com/clap-rs/clap/issues/568_

----

For some argument values, the bash-completions may want to include additional logic for what type of value to complete, rather than allowing arbitrary strings.  For instance, an option might accept a path to a file that exists; bash has a mechanism for that.  Or, an option might accept the name of a git ref that exists; that's something more easily implemented in the program in Rust.  Either way, it makes sense to augment clap's completion logic.

This also argues for implementing the completions by calling the program at runtime, rather than via shell; that way, arbitrary logic in Rust can determine what to complete, rather than providing a template system to feed in shell code (ick).

**Comment by epage** [comment]
<a href="https://github.com/kbknapp"><img src="https://avatars.githubusercontent.com/u/6942134?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [kbknapp](https://github.com/kbknapp)**
_Monday Jul 04, 2016 at 21:35 GMT_

----

Perhaps adding something like was discussed in in #376 where there is a "completer" function? I'm all for this, but figured it's also addable in a backwards compatible way once I had the base implementation complete.

I'm just not sure which would be the _best_ way to add this so I'm open to all ideas.

**Comment by epage** [comment]
<a href="https://github.com/joshtriplett"><img src="https://avatars.githubusercontent.com/u/162737?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [joshtriplett](https://github.com/joshtriplett)**
_Monday Jul 04, 2016 at 22:38 GMT_

----

I'm honestly not sure either.  I'm really hesitant to suggest inlining shell script snippets into Rust code as strings; I'd rather see those written in a separate shell file and included from there (not least of which to get the right filetype and highlighting).  bash (via `compgen`) and bash-completion (via functions in `/usr/share/bash-completion/bash_completion`) have some built-in helpers, and it'd be nice to support those for common cases like hostnames, users, and files (with glob patterns).  Someone might also want to write arbitrary shell code to enumerate argument values.  It'd also be nice to support using arbitrary Rust code by invoking the program.

I think what I'd suggest is that the `.completer` function should take an enum argument, where that enum has values like `User`, `File`, `FileGlob("*.ext")`, `BashFunc("__comp_function_name")`, and `RustFunc(...)`.  Those would then translate into appropriate calls to compgen, calls to the specified function, or invocations of the program to run Rust code.  (That last one would also require something like a `global_setting` to enable a `--clap-complete` option or similar.)

This is turning out to be a remarkably hairy yak.

**Comment by epage** [comment]
<a href="https://github.com/mathstuf"><img src="https://avatars.githubusercontent.com/u/97253?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [mathstuf](https://github.com/mathstuf)**
_Sunday Oct 02, 2016 at 12:15 GMT_

----

In zsh at least, clap could generate completion function calls such as:

``` zsh
(( $+functions[_appname_clap_complete_ARG] )) || _appname_clap_complete_ARG () {
}
```

Which can then be overridden in a supplemental file included before this one (via `source` if the file exists). Bash probably has some mechanism that works similarly.

**Comment by epage** [comment]
<a href="https://github.com/kbknapp"><img src="https://avatars.githubusercontent.com/u/6942134?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [kbknapp](https://github.com/kbknapp)**
_Monday Oct 10, 2016 at 19:15 GMT_

----

@emk Your post has me thinking about this more and more, I'm thinking some sort of hybrid between what @joshtriplett listed above and what you're proposing.

My schedule is pretty busy this week, but I should be able to at least test some ideas and see the feasibility. Stay tuned to this thread for updates!

**Comment by epage** [comment]
<a href="https://github.com/emk"><img src="https://avatars.githubusercontent.com/u/36963?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [emk](https://github.com/emk)**
_Monday Oct 10, 2016 at 23:43 GMT_

----

Another approach worth a quick glance might be optcomplete for Python:
http://furius.ca/optcomplete/ As far as I can tell, this uses one small,
universal shell script for each supported shell, and offloads all the
actual completion work to the application's own arg parsing machinery.

I think another Python library just uses a '--_complete' on the program
that does all the actual work, but I can't find it right now. I'll keep
Googling around when I have moment and post anything interesting I find.

Thank you so much for a great library and for looking into this!

**Comment by epage** [comment]
<a href="https://github.com/emk"><img src="https://avatars.githubusercontent.com/u/36963?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [emk](https://github.com/emk)**
_Tuesday Oct 11, 2016 at 10:23 GMT_

----

Ah, here we go. [Some docs](https://github.com/dbarnett/python-selfcompletion) on how several Python arg parsing libraries handle `--_completion`.

> selfcompletion is a layer on top of argparse to take the fine-grained model
> argparse builds of the arguments your program accepts and automatically
> generate an extra '--_completion' argument that generates all possible
> completions given a partial command line as a string.
> 
> The '--_completion' argument in turn is used by a generic bash programmable
> completion script that tries '--_completion' on any program that doesn't have
> its own completion already available, renders the output of the program's
> built-in completion if available, and otherwise silently falls back to the
> shell default.

Here is the [generic completion function for bash](https://github.com/dbarnett/python-selfcompletion/blob/master/bash_completion.d/_selfcompletion):

``` sh
_foo()
{
    prog="$1"
    while read glob_str; do
        case $prog in
        $glob_str)
            return 1;;
        esac
    done < <( echo "$SELFCOMPLETION_EXCLUSIONS" )
    which "$prog" >/dev/null || return 1
    _COMP_OUTPUTSTR="$( $prog --_completion "${COMP_WORDS[*]}" 2>/dev/null )"
    if test $? -ne 0; then
        return 1
    fi
    readarray -t COMPREPLY < <( echo -n "$_COMP_OUTPUTSTR" )
}

complete -o default -o nospace -D -F _foo
```

The advantage of this approach is that the per-shell code can be written only once, and all the hard work can be done directly by the application itself. Obviously, there might be disadvantages as well. But I figured it was worth tracking down all the existing attempts to standardize this to see if any of them had helpful ideas. :-)

**Comment by epage** [comment]
<a href="https://github.com/joshtriplett"><img src="https://avatars.githubusercontent.com/u/162737?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [joshtriplett](https://github.com/joshtriplett)**
_Monday Oct 24, 2016 at 21:40 GMT_

----

@kbknapp Any updates on this mechanism? I have someone asking after completions, and I'd love to beta-test this.

**Comment by epage** [comment]
<a href="https://github.com/jcreekmore"><img src="https://avatars.githubusercontent.com/u/385202?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [jcreekmore](https://github.com/jcreekmore)**
_Tuesday Oct 25, 2016 at 17:35 GMT_

----

@kbknapp I would be interested in this as well. I am currently post-processing my completions to substitute in `_filedir` for filename completion, but that is less than ideal.

**Comment by epage** [comment]
<a href="https://github.com/kbknapp"><img src="https://avatars.githubusercontent.com/u/6942134?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [kbknapp](https://github.com/kbknapp)**
_Tuesday Oct 25, 2016 at 17:53 GMT_

----

@joshtriplett @jcreekmore 

Now that the ZSH implementation is complete I've got a better handle on this. The biggest issue I see holding this up is that completions are done differently between all three (so far) supported shells.

I'm all for some sort of enum with variants that allow things like, `Files`, `Directories`, `Globs(&str)`, `Code(&str)` or something to that effect. But some shells support those things verbatim, others only in arbitrary ways that clap doesn't use when gen'ing the completion code.

I'm just unsure of the best way to expose this. Perhaps on an `Arg::complete_with(enum)`?

I guess, first what is the shell you're trying to support, and what particular portions are you wanting to inject into the completion code?

**Comment by epage** [comment]
<a href="https://github.com/emk"><img src="https://avatars.githubusercontent.com/u/36963?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [emk](https://github.com/emk)**
_Tuesday Oct 25, 2016 at 18:00 GMT_

----

@kbknapp For `cage`, we'd like to be able to complete custom "types" of values, such as Docker container names, "pod" names, target environments, and so on. The legal values can only be determined by asking our executable at runtime, since they vary from project to project.

This is pretty much how `git-completion` handles origin names, branch names, etc.

The problem with an `enum` is that it would limit us to just a few built-in types such as `Files`, `Directories`, etc., right?

**Comment by epage** [comment]
<a href="https://github.com/joshtriplett"><img src="https://avatars.githubusercontent.com/u/162737?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [joshtriplett](https://github.com/joshtriplett)**
_Tuesday Oct 25, 2016 at 19:23 GMT_

----

@kbknapp I don't think you need to support embedding arbitrary shell code from Rust.  My suggestion would be to support the lowest-common-denominator bits (filenames, usernames, filenames matching one of these patterns, etc), and then have a "call this Rust function" variant that invokes the program itself with some internal hidden --clap-complete option that dispatches to that Rust function.  That makes it easy to do things like "a git ref matching this pattern", by calling a Rust function implementing that.

For those common categories like filenames or usernames, use the shell built-in mechanisms if available, or provide and use code implementing those simple completions if the shell doesn't have them.

If people want "invoke this shell code", I'd suggest adding a Rust variant to call a named shell function, and then letting people add that shell function to the resulting generated completions for any shell they support.  That seems preferable to embedding many variants of shell code directly.

``` rust
enum Completion<F: Fn(...) -> ...> {
    File,
    User,
    Hostname,
    Strings(&[str]),
    Globs(&[str]),
    ShellFunction(&str), // maybe
    RustFunction(F),
}
```

**Comment by epage** [comment]
<a href="https://github.com/kbknapp"><img src="https://avatars.githubusercontent.com/u/6942134?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [kbknapp](https://github.com/kbknapp)**
_Tuesday Oct 25, 2016 at 19:36 GMT_

----

@emk Yes, and no. It would be extensible, so more variants could be added. But Some of the variants could also take additional parameters, and ultimately (possibly) injecting arbitrary shell script via something like ,`Code(&str)` which of course isn't _super_ great, but perhaps a fallback if a particular variant doesn't quite fit the bill. (Or perhaps not...it could end up being massively unsafe :stuck_out_tongue_winking_eye: )

At the same time, I haven't looked into _exactly_ where this code would be injected and ultimately if it's even feasible yet. This is just straight of the top of my head right now.

Also, if the "types" are know prior to runtime ZSH already supports this just by using the `Arg::possible_values`

**Comment by epage** [comment]
<a href="https://github.com/joshtriplett"><img src="https://avatars.githubusercontent.com/u/162737?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [joshtriplett](https://github.com/joshtriplett)**
_Tuesday Oct 25, 2016 at 19:54 GMT_

----

That's true, "one of these fixed strings" should be an option as well.  Updating the type in my previous comment.

**Comment by epage** [comment]
<a href="https://github.com/kbknapp"><img src="https://avatars.githubusercontent.com/u/6942134?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [kbknapp](https://github.com/kbknapp)**
_Saturday Jan 14, 2017 at 02:25 GMT_

----

From @Xion in #816 

> In Python, there is a package called [argcomplete](https://pypi.python.org/pypi/argcomplete) which provides very flexible autocompletion for apps that use the standard argparse module. What it allows is to implement a custom completion provider: essentially a piece of your own code that's executed when the the binary is invoked in a Special Way (tm) by the shell-specific completion script.
> For an example, see [here](https://github.com/Xion/gisht.py/blob/master/gisht/args/autocomplete.py#L26). The code is preparing completions dynamically from the filesystem, or even from a remote API (if certain flag isn't passed (flags are partially parsed at this point)).
> Having something like this in clap would be very nice. I know this is a potentially complex subsystem so it'd be unreasonable to expect it implemented anytime, but I wanted to at least put this feature on the radar.

**Comment by epage** [comment]
<a href="https://github.com/kbknapp"><img src="https://avatars.githubusercontent.com/u/6942134?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [kbknapp](https://github.com/kbknapp)**
_Saturday Jan 14, 2017 at 02:48 GMT_

----

After reading through some of the `argcomplete` python module the hardest part will be figuring out how to call a Rust function from the shell.

**Comment by epage** [comment]
<a href="https://github.com/kbknapp"><img src="https://avatars.githubusercontent.com/u/6942134?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [kbknapp](https://github.com/kbknapp)**
_Saturday Jan 14, 2017 at 02:52 GMT_

----

I'm guessing what'll end up happening is some sort of double run with hidden args.

**Comment by epage** [comment]
<a href="https://github.com/kbknapp"><img src="https://avatars.githubusercontent.com/u/6942134?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [kbknapp](https://github.com/kbknapp)**
_Thursday Jan 19, 2017 at 01:30 GMT_

----

Expanding on the ideas (from #818)

> The problem with implementing this is I just haven't had a good time to sit down and think about how (because of work, holidays, family, etc.). I want a way to specify this that abstracts well enough to work for all shells. The easiest way is to say, "Put your arbitrary completion shell script here inside this `Arg::complete_with(&str)`" but that feels super hacky to me, and potentially unsafe. What I'd like to do is provide a `Arg::complete_with(Fn(&str, &str)->String)` (and a `Arg::complete_with_os(Fn(&str,&OsStr)->OsString)` where an arbitrary Rust function is called...but herein lies the problem; shell completions are run before the program executes. This has led to some people using hidden args or something like, `$ prog --complete me<tab>` calling a shell completer that actually runs `$ prog --_complete_arg "complete" --_complete_prefix "me"` which generates the possible completions and returns them to the shell. I'm not against doing that, but again feels strange because you're injecting hidden args into a CLI. Although typing this out right now does make me lean towards this solution.

**Comment by epage** [comment]
<a href="https://github.com/joshtriplett"><img src="https://avatars.githubusercontent.com/u/162737?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [joshtriplett](https://github.com/joshtriplett)**
_Thursday Jan 19, 2017 at 06:21 GMT_

----

@kbknapp I like the idea of using parameters like `--_clap-complete` and similar, and then calling the program itself to do the completion via Rust code.  That would make it possible to (for instance) use git2-rs from Rust to complete names of things in a git repository.

Naming the argument/parameter seems sufficient to dispatch to the right function, though I'd also like to have the other arguments available to handle things like `prog --repository /path/to/.git --branch someth[tab]` (which needs the repository to complete branches from).

Also, those `complete_with` functions should return either a `Vec<OsString>` or in general an implementation of `Iterator<OsString>`, to return all completions.  (The latter would benefit from but not require `-> impl Trait` support, since `complete_with` can declare the return value as a generic while the actual function/closure might use a specific iterator type.)

And what does the first `&str` parameter of those functions refer to?

**Comment by epage** [comment]
<a href="https://github.com/mathstuf"><img src="https://avatars.githubusercontent.com/u/97253?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [mathstuf](https://github.com/mathstuf)**
_Thursday Jan 19, 2017 at 14:29 GMT_

----

I'd just like to note that for really expensive queries (imagine tab-completion for `cargo install <Tab>`), at least `zsh` supports a cache for completions; it would be nice to have a way to leverage that cache via some "can be cached" flag (cache expiry is controlled via `zstyle`).

**Comment by epage** [comment]
<a href="https://github.com/joshtriplett"><img src="https://avatars.githubusercontent.com/u/162737?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [joshtriplett](https://github.com/joshtriplett)**
_Thursday Jan 19, 2017 at 20:05 GMT_

----

> I'd just like to note that for really expensive queries (imagine tab-completion for `cargo install <Tab>`)

I'd expect tab-completion for `cargo install <Tab>` to list all the currently-available crates, without doing a network update first.  That would use entirely local information, and generate completions quickly.

Rather than attempting to integrate with any particular shell's completion caching, perhaps if an app considers its completions expensive to generate, it could cache the necessary information to make them fast?  Caching policies seem much easier to implement in Rust, since they could use arbitrary freshness metrics ("the underlying data hasn't changed so return the cached list").

**Comment by epage** [comment]
<a href="https://github.com/kbknapp"><img src="https://avatars.githubusercontent.com/u/6942134?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [kbknapp](https://github.com/kbknapp)**
_Thursday Jan 19, 2017 at 21:31 GMT_

----

@mathstuf @joshtriplett yeah, I'd prefer not to incorporate shell specific arguments if at all possible and leave that to the Rust function of the implementer.

> And what does the first `&str` parameter of those functions refer to?

It referred to the current arg being parsed (as determined by `clap`), the second was the prefix being completed (if any). Whether it's a `Vec<String>` or `String` (including `\n`s) (or `OsString` equivalents) returned will be determined once I'm able look at all four shells and see what they're expecting.

> I'd also like to have the other arguments available

I thought about this as well, and I'm torn between providing a simple list of strings (at which point why include them at all because people could just use `std::env::args`), or some magic about "allowing failed/incomplete parses" to give off a `ArgMatches` struct which I'd assume is actually the information you'd want but is more complex to implement.

**Comment by epage** [comment]
<a href="https://github.com/joshtriplett"><img src="https://avatars.githubusercontent.com/u/162737?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [joshtriplett](https://github.com/joshtriplett)**
_Thursday Jan 19, 2017 at 22:57 GMT_

----

@kbknapp Parsing partial arguments (without enforcing all of the argument requirements) seems like a pain, but I'd rather not manually implement argument parsing in order to do completion.  That said, implementing this without support for parsing other arguments at first would still help in many cases.

Does the "current arg being parsed" exist to allow passing the same completion function for multiple arguments, and distinguishing them via argument?  If so, why not just do that using a closure?  You could pass `|arg| complete("foo", arg)` easily enough.  Completing multiple arguments with the same function but needing to distinguish between them by name seems sufficiently uncommon to not want to complicate the common case.

**Comment by epage** [comment]
<a href="https://github.com/kbknapp"><img src="https://avatars.githubusercontent.com/u/6942134?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [kbknapp](https://github.com/kbknapp)**
_Saturday Jan 21, 2017 at 22:18 GMT_

----

>  Parsing partial arguments (without enforcing all of the argument requirements) seems like a pain

Actually I don't think it will be. `clap` enforces all the requirements lazily, so a single branch would allow a failing parse to "pass" and only in that very strict circumstance. Of course this would be it'd have to be well documented that when using the `ArgMatches` given to the completing function, those requirements haven't been enforced yet and can't be relied on. But it would at least allow one to check for the presence of args, values, etc.

> If so, why not just do that using a closure?

Actually I like that idea, I'll have to try this out when I get some time to sit down and try implementing this!

**Comment by epage** [comment]
<a href="https://github.com/droundy"><img src="https://avatars.githubusercontent.com/u/137513?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [droundy](https://github.com/droundy)**
_Friday Aug 11, 2017 at 12:50 GMT_

----

I would suggest rather than an enum that multiple methods setting completion possibilities would be a better and more extensible API.  So rather than

    Args::complete_with(enum)

you would have

    Args::complete_with_files()

and a whole set of other `complete_with_XXX` methods.  This removes the requirement that you foresee every possibility on the first version of the API (since you can't add variants to a public enum without breaking backwards compatibility).  I think the most important one is the one that uses an auto-generated flag to call rust code, since this can then implement all the others in a shell-independent way.  Then optimizations could be made to do the others in a shell-dependent way if that seems to help.  So I would focus on something like:

    Args::complete_with_function(|matches_so_far| -> [String] { ... }) 

where it is important to provide the completed flags, since that can affect what is a valid completion for a given flag.

**Comment by epage** [comment]
<a href="https://github.com/droundy"><img src="https://avatars.githubusercontent.com/u/137513?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [droundy](https://github.com/droundy)**
_Friday Aug 11, 2017 at 12:59 GMT_

----

Here is an example of a bash completion that simply calls command-line flags that return the possible completions for [darcs](https://github.com/kbknapp/clap-rs/files/1217871/darcs.txt).  darcs does no shell-specific munging in its Haskell code (since it is usable by multiple shells), just outputs a `'\n'`-delimited list of completions.  Then the bash code handles spaces and colons specially for bash.

**Comment by epage** [comment]
<a href="https://github.com/softprops"><img src="https://avatars.githubusercontent.com/u/2242?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [softprops](https://github.com/softprops)**
_Tuesday Jun 05, 2018 at 03:20 GMT_

----

Here's some inspiration from how golang kingpin package handles dynamic tab completion using the bash [COMPREPLY bash protocol](https://www.gnu.org/software/bash/manual/html_node/A-Programmable-Completion-Example.html)

Kingpin supports dynamic completions via it's [HintAction interface](https://github.com/alecthomas/kingpin/blob/master/README.md#additional-api)

I imagine this should be possible providing a fn interface to clap's Arg type for dynamic completions

Under the covers kingpin forks program control when in [completion mode](https://github.com/alecthomas/kingpin/blob/master/app.go#L205) to invoke that completion func then exits the process.

It switches modes of operation based on a [flag](https://github.com/alecthomas/kingpin/blob/master/app.go#L66) that the completion script [passes](https://github.com/alecthomas/kingpin/blob/master/templates.go#L240).

**Comment by epage** [comment]
<a href="https://github.com/adamtulinius"><img src="https://avatars.githubusercontent.com/u/749381?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [adamtulinius](https://github.com/adamtulinius)**
_Thursday Sep 06, 2018 at 09:19 GMT_

----

> Here's some inspiration from how golang kingpin package handles [..]

Please not that kingpin currently can't complete file paths, which need something like https://github.com/alecthomas/kingpin/compare/master...DBCDK:hint-files to fix. I'm just mentioning this, because it required work on the compreply stuff, and might be useful here as well.
